mono/metadata/threads.c

```diff
@@ -464,35 +502,122 @@ static void ensure_synch_cs_set (MonoInternalThread *thread)
+
+	if (ret != NULL) {
 		/* Another thread must have installed this CS */
+		/* Or the finalizer had to have run */
 		mono_coop_mutex_destroy (synch_cs);
 		g_free (synch_cs);
 	}
 }
 
-static inline void
-lock_thread (MonoInternalThread *thread)
+static void
+internal_thread_free (MonoInternalThread *thread)
 {
+	MonoCoopMutex *synch_cs = thread->synch_cs;
+	thread->synch_cs = INTERNAL_THREAD_TOMBSTONE;
+	mono_memory_barrier ();
+
+	mono_coop_mutex_destroy (synch_cs);
+	g_free (synch_cs);
+}
+
+static void
+cleanup_thread_lock (MonoInternalThread *thread)
+{
+	thread->freeing = 1;
+	mono_memory_barrier ();
+
+	gint32 waiters = mono_atomic_load_i32 (&thread->num_waiters);
+
+	// If we started cleanup without any writers, finish the freeing
+	// ourselves. Else the thread who unlocks the thread after the last pending 
+	// wait is finished will do the freeing.
+	if (waiters == 0)
+		internal_thread_free (thread);
+}
+
+static void
+decrement_wait_queue (MonoInternalThread *thread)
+{
+	gint32 waiters = mono_atomic_dec_i32 (&thread->num_waiters);
+	if (waiters == 0) {
+		gboolean freeing = mono_atomic_load_i32 (&thread->freeing);
+		if (freeing)
+			internal_thread_free (thread);
+	}
+}
+
+static void
+increment_wait_queue (MonoInternalThread *thread)
+{
+	// We will have a pointer to the lock in our registers, need to warn
+	// finalizer to wait for us to finish.
+	mono_atomic_inc_i32 (&thread->num_waiters);
+}
+
+static void
+wait_on_lock (MonoInternalThread *thread, MonoCoopMutex *lock)
+{
+	mono_coop_mutex_lock (lock);
+	decrement_wait_queue (thread);
+}
+
+static MonoCoopMutex *
+get_lock (MonoInternalThread *thread, gboolean assert_fine)
+{
+	increment_wait_queue (thread);
+
 	if (!thread->synch_cs)
 		ensure_synch_cs_set (thread);
 
-	g_assert (thread->synch_cs);
+	gpointer lock = mono_atomic_load_ptr ((volatile gpointer *) &thread->synch_cs);
+	g_assert (lock);
+
+	gboolean freeing = mono_atomic_load_i32 (&thread->freeing);
+
+	if (G_UNLIKELY(freeing)) {
+		if (assert_fine)
+			g_error ("Attempting to lock a thread that has been finalized.");
+
+		// We lied, we actually are not going to be joining the queue
+		// of threads that have a reference to the lock and will attempt to
+		// use it.
+		decrement_wait_queue (thread);
+		return NULL;
+	}
+	return (MonoCoopMutex *) lock;
+}
+
+static gboolean
+try_lock_thread (MonoInternalThread *thread)
+{
+	MonoCoopMutex *lock = get_lock (thread, FALSE);
+	if (lock)
+		wait_on_lock (thread, lock);
+
+	return lock != NULL;
+}
 
-	mono_coop_mutex_lock (thread->synch_cs);
+static void
+assert_lock_thread (MonoInternalThread *thread)
+{
+	wait_on_lock (thread, get_lock (thread, FALSE));
 }
 
-static inline void
+static void
 unlock_thread (MonoInternalThread *thread)
 {
-	mono_coop_mutex_unlock (thread->synch_cs);
+	gpointer lock = mono_atomic_load_ptr ((volatile gpointer *) &thread->synch_cs);
+	if (G_LIKELY (lock != INTERNAL_THREAD_TOMBSTONE))
```

Comment:
This state doesn't occur in my proposal either. Every lock must be balanced by an unlock. The "waiter" count is kept incremented until unlock, rather than where you decrement it right after lock.

Reply:
I was attempting to minimize visible changes in threading behavior. You're right though, changing a refcount at unlock time is more in alignment with expected behavior

Reply:
I'm going for correctness here.